### ISSUE TITLE
What are tf.IndexedSlices and dense gradients? Why do a keras optimizer distinguish between them?

### ISSUE THREAD

**Issue by MidouDali** [keras-team/keras#17500]
I want to write a custom optimizer with Keras. Should I distinguish between tf.IndexedSlices and dense gradients?
I  want to know what is the difference between tf.IndexedSlices and dense gradients. Why do a keras optimizer distinguish between them while updating the variables at each step? thank you

**Comment by tilakrayal** [comment]
@MidouDali,
**tf.IndexedSlices** is the sparse representation of a set of tensor slices at given indices. An IndexedSlices is typically used to represent a subset of a larger tensor dense of shape and the values in indices are the indices in the first dimension of the slices that have been extracted from the larger tensor.

**tf.IndexedSlices(
    values, indices, dense_shape=None
)**

The IndexedSlices class is used principally in the definition of gradients for operations that have sparse gradients

```
v = tf.Variable([[0.,1, 2], [2, 3, 4], [4, 5, 6], [6, 7, 8]])
with tf.GradientTape() as tape:
  r = tf.gather(v, [1,3])
index_slices = tape.gradient(r,v)
index_slices.indices.numpy()
```

**tf.keras.optimizers.Optimizer** is abstract optimizer base class. If we want to implement your own optimizer, please subclass this class instead of **_BaseOptimizer**.
And when we want to process the gradients before applying them you can instead use the optimizer in three steps:

1. Compute the gradients with **tf.GradientTape.**
2. Process the gradients as you wish.
3. Apply the processed gradients with **apply_gradients()**.

**Comment by MidouDali** [comment]
Hi,
thank you
All existing optimizers in keras (adam, sgd, nadam, ...) distinguish between  tf.IndexedSlices gradients and dense gradients in order to update variables (in update_step() function):
def update_step(self, gradient, variable):
      if isinstance(gradient, tf.IndexedSlices):
              update manner 1
      else:# Dense gradients.
              update manner 2 (other update manner)

I want to write a custom optimizer and I want why should I distinguish between tf.IndexedSlices gradients and dense gradients?
thank you

**Comment by MidouDali** [comment]
any help please!

**Comment by tilakrayal** [comment]
If you want to create your own optimization algorithm, please try to inherit from the class and override the below methods:

build: Create your optimizer-related variables, such as momentums in SGD optimizer.
update_step: Implement your optimizer's updating logic.
get_config: serialization of the optimizer, include all hyper parameters.

Your optimizer would automatically be compatible with tensorflow distributed training if you subclass optimizer_experimental.Optimizer. Also please take a look at the official document [link](https://www.tensorflow.org/api_docs/python/tf/keras/optimizers/Optimizer#methods_2) and try to follow the methods mentioned. Thank you!
